File: 061_rotate_list/Solution.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
    }
}

class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) return null;
        ListNode tail = head;
        int length = 1;
        while (tail.next != null) {
            length++;
            tail = tail.next;
        }
        tail.next = head;
        k = length - k % length;
        while (k-- > 0) tail = tail.next;
        head = tail.next;
        tail.next = null;
        return head;
    }

    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " ");
            head = head.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode tmp = head;
        for (int i = 2; i <= 5; i++) {
            tmp.next = new ListNode(i);
            tmp = tmp.next;
        }
        Solution solution = new Solution();
        head = solution.rotateRight(head, 11);
        printList(head);
    }
}

File: 061_rotate_list/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head: return None
        tail = head
        length = 1
        while tail.next:
            length += 1
            tail = tail.next
        tail.next = head
        k = length - k % length
        for _ in range(k, 0, -1): tail = tail.next
        head = tail.next
        tail.next = None

        return head


def print_list(head: ListNode):
    while head:
        print(head.val, end = ' ')
        head = head.next


if __name__ == '__main__':
    head = ListNode(1)
    tmp = head
    for i in range(2, 6):
        tmp.next = ListNode(i)
        tmp = tmp.next
    head = Solution().rotateRight(head, 12)
    print_list(head)
